transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,636 @@
+// Copyright 2017-2020 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+//
+use crate::{error::Error, socket, verifier, Upgrade};
+use async_macros::ready;
+use async_std::{net::SocketAddr, task::spawn};
+use futures::{channel::mpsc, prelude::*};
+use libp2p_core::{
+    multiaddr::{host_addresses, Multiaddr, Protocol},
+    transport::{ListenerEvent, TransportError},
+    Transport,
+};
+use log::{debug, info, trace, warn};
+use parking_lot::Mutex;
+use quinn_proto::{Connection, ConnectionHandle};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    sync::{Arc, Weak},
+    task::{Context, Poll},
+    time::Instant,
+};
+
+/// Represents the configuration for a QUIC/UDP/IP transport capability for libp2p.
+///
+/// The QUIC endpoints created by libp2p will need to be progressed by running the futures and streams
+/// obtained by libp2p through a reactor.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The client configuration.  Quinn provides functions for making one.
+    pub client_config: quinn_proto::ClientConfig,
```

Comment:
I'd prefer to not expose anything related to `quinn` in the public API.
This should either be private or removed. If some options are important, then let's wrap around the `quinn` config rather than exposing that config directly.

Same remark for the two other fields.

Reply:
Done.  Some tunables should be exposed at some point, but those can wait.